### ISSUE TITLE
`non_topologically_sorted_functions` can produce unsatisfiable ordering constraints

### ISSUE THREAD

**Issue by smoelius** [trailofbits/dylint#2012]
Codex-generated description appears below.

My gut says this should be the right fix:
> prioritize caller-before-callee constraints over sibling-order constraints

But I would need to investigate whether that completely resolves the problem.

## Summary

The [`non_topologically_sorted_functions` lint][lint-readme] can produce a cycle of
ordering constraints for an acyclic call graph. In that situation, repeatedly applying
the lint's suggested function moves never reaches a fixpoint: fixing one warning exposes
another, and no ordering of the functions can satisfy all constraints.

The problem occurs because the lint requires both:

1. callers to appear before callees; and
2. callees to appear in the order in which a common caller first invokes them.

The second rule can conflict transitively with the first. The implementation avoids only
an immediately reversed constraint; it does not detect cycles in the complete constraint
graph.

## Minimal reproducer

```rust
fn root() {
    shared();
    branch();
}

fn shared() {}

fn branch() {
    middle();
}

fn middle() {
    shared();
}
```

The call graph is acyclic:

```text
root ──→ shared
  │
  └────→ branch ──→ middle ──→ shared
```

However, the lint constructs these ordering constraints:

```text
shared < branch   // sibling call order in root
branch < middle   // caller before callee
middle < shared   // caller before callee
```

Together, they form an unsatisfiable cycle:

```text
shared < branch < middle < shared
```

There is therefore no source ordering that can satisfy the lint.

## Actual behavior

The lint reports one or more functions as out of order and recommends moving them earlier.
Following those recommendations eventually returns to a previously violated constraint.

In [`cargo-unmaintained`][cargo-unmaintained], the relevant structure is:

```text
is_unmaintained_package
├── clone_repository
├── outdated_deps
└── latest_commit_age
    └── timestamp
        ├── clone_repository
        └── timestamp_uncached
            └── timestamp_from_clone
                └── clone_repository
```

The direct call to [`clone_repository` from `is_unmaintained_package`][direct-clone-call]
places it before [`latest_commit_age`][latest-commit-call] under the sibling-order rule.
But the timestamp path eventually calls [`clone_repository` from
`timestamp_from_clone`][nested-clone-call], requiring it to appear later.

Repeatedly applying the diagnostics produced this sequence:

1. `clone_repository` and `outdated_deps` should precede `latest_commit_age`;
2. `latest_commit_age` should precede `timestamp`;
3. `timestamp_from_clone` should precede `clone_repository`;
4. continuing the moves necessarily returns to an earlier constraint.

A crate-level allowance was ultimately required even though the underlying call graph is
not recursive.

## Expected behavior

The lint should not suggest a sequence of moves when its constraint graph is cyclic.
Possible behaviors include:

- detect strongly connected components in the ordering-constraint graph and omit
  contradictory sibling-order constraints;
- prioritize caller-before-callee constraints over sibling-order constraints;
- add a sibling-order constraint only if doing so would not create a cycle; or
- emit one diagnostic explaining that the requested ordering is unsatisfiable.

## Implementation notes

The lint explicitly creates caller-before-callee constraints in
[`build_caller_callee_constraint`][caller-callee].

It creates pairwise sibling-order constraints in
[`build_callee_order_constraints`][callee-order]. Before inserting an edge, it checks only
whether the exact reverse edge already exists:

```rust
if must_come_before.contains(&(b, a)) {
    continue;
}
```

That check prevents a two-node cycle but not a longer cycle such as
`a → b → c → a`.

Finally, [`find_violations`][find-violations] compares each constraint independently
against source positions. There is no cycle detection or topological-sort attempt before
diagnostics are emitted.

## Environment

- Dylint commit: [`1e431b6c9181ee78a24c3ad3f08ad95d4bbff2e8`][dylint-commit]
- Reproducing project: `cargo-unmaintained`
- Command: `cargo dylint --all`

[lint-readme]: https://github.com/trailofbits/dylint/tree/1e431b6c9181ee78a24c3ad3f08ad95d4bbff2e8/examples/restriction/non_topologically_sorted_functions
[dylint-commit]: https://github.com/trailofbits/dylint/commit/1e431b6c9181ee78a24c3ad3f08ad95d4bbff2e8
[caller-callee]: https://github.com/trailofbits/dylint/blob/1e431b6c9181ee78a24c3ad3f08ad95d4bbff2e8/examples/restriction/non_topologically_sorted_functions/src/lib.rs#L120-L144
[callee-order]: https://github.com/trailofbits/dylint/blob/1e431b6c9181ee78a24c3ad3f08ad95d4bbff2e8/examples/restriction/non_topologically_sorted_functions/src/lib.rs#L146-L172
[find-violations]: https://github.com/trailofbits/dylint/blob/1e431b6c9181ee78a24c3ad3f08ad95d4bbff2e8/examples/restriction/non_topologically_sorted_functions/src/lib.rs#L174-L223
[cargo-unmaintained]: https://github.com/trailofbits/cargo-unmaintained/blob/4c36f809bdac836e4c5f89d9acbb9f465545481e/src/lib.rs
[direct-clone-call]: https://github.com/trailofbits/cargo-unmaintained/blob/4c36f809bdac836e4c5f89d9acbb9f465545481e/src/lib.rs#L684
[latest-commit-call]: https://github.com/trailofbits/cargo-unmaintained/blob/4c36f809bdac836e4c5f89d9acbb9f465545481e/src/lib.rs#L705
[nested-clone-call]: https://github.com/trailofbits/cargo-unmaintained/blob/4c36f809bdac836e4c5f89d9acbb9f465545481e/src/lib.rs#L977

**Comment by smoelius** [comment]
Fixed by #2032
